File: FinalProject/server/book_smart/models/listing.py
"""
listing.py

Created by Stephen Andrews, February 19th, 2018.
"""

from book_smart.extensions import db

listing_identifier = db.Table('listing_identifier',
    db.Column('listing_id', db.Integer, db.ForeignKey('listing.listing_id')),
    db.Column('listing_type_id', db.Integer, db.ForeignKey('listing_type.listing_type_id'))
)

class ListingType(db.Model):

    listing_type_id = db.Column(db.Integer, primary_key=True)
    listing_type = db.Column(db.Text, nullable=False)


class Listing(db.Model):
    """Represents one of three listings -- sale, loan, or swap."""

    listing_id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.Text, db.ForeignKey('user.username'), nullable=False)
    isbn = db.Column(db.Text, db.ForeignKey('book.isbn'), nullable=False)
    condition = db.Column(db.Text, nullable=False)
    price = db.Column(db.Float)
    listing_types = db.relationship('ListingType', secondary=listing_identifier)

    def to_json(self):
        listing = {}
        listing['listingId'] = self.listing_id
        listing['username'] = self.username
        listing['isbn'] = self.isbn
        listing['condition'] = self.condition
        listing['price'] = self.price
        listing['listingTypes'] = [l.listing_type for l in self.listing_types]

        return listing

File: FinalProject/WPIReads/app/src/main/java/cs4518_team6/booksmart/model/Listing.java
package cs4518_team6.booksmart.model;

import android.util.Log;

import com.github.kevinsawicki.http.HttpRequest;
import com.google.gson.Gson;

import java.net.HttpURLConnection;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cs4518_team6.booksmart.Constants;

/**
 * Represents a book being available for purchasing, loaning, or swapping.
 * Created by Stephen Andrews.
 */
public class Listing extends Model {

    private static final String ENDPOINT = "/listings/";

    /**
     * Represents the available listing types.
     */
    public enum ListingType {
        FOR_TRADE,
        FOR_LOAN,
        FOR_SALE
    }

    private Integer listingId;
    private String username;
    private String isbn;
    private String condition;
    private Double price;
    private String[] listingTypes;

    public static boolean add(String username, String isbn, String condition, Double price, final List<ListingType> listingTypes) {
        String url = Constants.API_URL + ENDPOINT + "add";
        Map<String, String> data = new HashMap<>();
        data.put("username", username);
        data.put("isbn", isbn);
        data.put("condition", condition);
        if (price != null) {
            data.put("price", String.valueOf(price));
        }

        List<Integer> listingTypeIds = new ArrayList<>();
        for (ListingType l : listingTypes) {
            listingTypeIds.add(l.ordinal() + 1);
        }
        data.put("listingTypes", listingTypeIds.toString());

        return HttpRequest.post(url).form(data).code() == HttpURLConnection.HTTP_OK;
    }

    /**
     * Delete a listing.
     * @param listingId The id of the listing to delete.
     */
    public static boolean delete(int listingId) {
        String url = Constants.API_URL + ENDPOINT + "delete/" + listingId;

        return HttpRequest.delete(url).code() == HttpURLConnection.HTTP_OK;
    }

    public static List<Listing> getAll() {
        String url = Constants.API_URL + ENDPOINT;
        HttpRequest request = HttpRequest.get(url);
        String response = request.body();

        return Arrays.asList(new Gson().fromJson(response, Listing[].class));

    }

    public Integer getListingId() {
        return listingId;
    }

    public String getUsername() {
        return username;
    }

    public String getIsbn() {
        return isbn;
    }

    public String getCondition() {
        return condition;
    }

    public Double getPrice() {
        return price;
    }

    public String[] getListingTypes() {
        return listingTypes;
    }

    @Override
    public String toString() {
        return "Listing{" +
                "listingId=" + listingId +
                ", username='" + username + '\'' +
                ", isbn='" + isbn + '\'' +
                ", condition='" + condition + '\'' +
                ", price=" + price +
                ", listingTypes=" + Arrays.toString(listingTypes) +
                '}';
    }
}
